components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -52,39 +182,857 @@ impl ReadableByteStreamController {
+            total_bytes_to_copy_remaining -= bytes_to_copy;
+        }
+
+        // If ready is false,
+        if !ready {
+            // Assert: controller.[[queueTotalSize]] is 0.
+            assert!(self.queue_total_size.get() == 0.0);
+
+            // Assert: pullIntoDescriptor’s bytes filled > 0.
+            assert!(pull_into_descriptor.bytes_filled() > 0);
+
+            // Assert: pullIntoDescriptor’s bytes filled < pullIntoDescriptor’s minimum fill.
+            assert!(pull_into_descriptor.bytes_filled() < pull_into_descriptor.minimun_fill);
+        }
+
+        // Return ready.
+        ready
+    }
+
+    /// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-fill-head-pull-into-descriptor>
+    pub(crate) fn fill_head_pull_into_descriptor(
+        &self,
+        bytes_copied: usize,
+        pull_into_descriptor: &PullIntoDescriptor,
+    ) {
+        // Assert: either controller.[[pendingPullIntos]] is empty,
+        // or controller.[[pendingPullIntos]][0] is pullIntoDescriptor.
+        assert!(
+            self.pending_pull_intos.borrow().is_empty() ||
+                self.pending_pull_intos.borrow().first().unwrap() == pull_into_descriptor
+        );
+
+        // Assert: controller.[[byobRequest]] is null.
+        assert!(self.byob_request.get().is_none());
+
+        // Set pullIntoDescriptor’s bytes filled to bytes filled + size.
+        pull_into_descriptor.set_bytes_filled(pull_into_descriptor.bytes_filled() + bytes_copied);
+    }
+
+    /// <https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamcontrollerenqueuedetachedpullintotoqueue>
+    pub(crate) fn enqueue_detached_pull_into_to_queue(
+        &self,
+        first_descriptor: &PullIntoDescriptor,
+    ) {
+        // Assert: pullIntoDescriptor’s reader type is "none".
+        assert!(first_descriptor.reader_type.is_none());
+
+        // If pullIntoDescriptor’s bytes filled > 0, perform ?
+        // ReadableByteStreamControllerEnqueueClonedChunkToQueue(controller,
+        // pullIntoDescriptor’s buffer, pullIntoDescriptor’s byte offset, pullIntoDescriptor’s bytes filled).
+
+        if first_descriptor.bytes_filled() > 0 {
+            self.enqueue_cloned_chunk_to_queue(
+                &first_descriptor.buffer.borrow(),
+                first_descriptor.byte_offset,
+                first_descriptor.bytes_filled(),
+            );
+        }
+
+        // Perform ! ReadableByteStreamControllerShiftPendingPullInto(controller).
+        self.shift_pending_pull_into();
+    }
+
+    /// <https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamcontrollerenqueueclonedchunktoqueue>
+    #[allow(unsafe_code)]
+    pub(crate) fn enqueue_cloned_chunk_to_queue(
+        &self,
+        buffer: &HeapBufferSource<ArrayBufferU8>,
+        byte_offset: usize,
+        byte_length: usize,
+    ) {
+        let cx = GlobalScope::get_cx();
```

Comment:
In https://github.com/servo/servo/pull/34844/commits/8eb060123373eb2b06af02ed93cd17ce73f69111 I tried to pass contexts and globals as much as possible by reference.